apache/incubator-zeppelin
# [ZEPPELIN-689] Add AngularJS z object and z.angularBind()

### What is this PR for?
Add client-side `z` object with method `angularBind()`

@Leemoonsoo 

Compared to the original implementation, I have simplified a lot.

Now, you can only bind angular variable to one unique scope, which is the **paragraph**. I just remove the note scope and also remove the `interpreter` parameter. 

Indeed, when passing a `paragraphId`, on the server-side, we can retrieve the `Paragraph` object with the `noteId` + `paragraphId` so we can now which interpreter is currently being used.

The signature of the `angularBind(varName, value, paragraphId)` method has also been greatly simplified.


_This is a sub-task of epic **[ZEPPELIN-635]**_
### What type of PR is it?
[Improvement]

### Todos

### Is there a relevant Jira issue?
**[ZEPPELIN-689]**

### How should this be tested?
* `git fetch origin pull/740/head:AngularJSBind`
* `git checkout AngularJSBind`
* `mvn clean package -DskipTests`
* `bin/zeppelin-daemon.sh restart`
* Create a new note
* In the first paragraph, put the following code

```html
%angular

<form class="form-inline">
  <div class="form-group">
    <label for="superheroId">Super Hero: </label>
    <input type="text" class="form-control" id="superheroId" placeholder="Superhero name ..." ng-model="superhero"></input>
  </div>
  <button type="submit" class="btn btn-primary" 
    ng-click="z.angularBind('superhero', superhero, 'PUT_HERE_PARAGRAPH_ID')"> Bind Angular</button>
</form>
```
* Create a second paragraph with the following code:

```scala
z.angular("superhero")
````

* Retrieve the paragraphId of the second paragraph
* In the first paragraph, replace the text **PUT_HERE_PARAGRAPH_ID** by the correct paragraph id
* In the input text, put "Superman" and click on the **Bind Angular** button
* Execute the second paragraph to see that the _superhero_ variable is now set to **Superman**

### Screenshots (if appropriate)
![angularbind](https://cloud.githubusercontent.com/assets/1532977/13251894/3e5324aa-da33-11e5-89a9-8702cc4c08f6.gif)

### Questions:
* Does the licenses files need update? --> **No**
* Is there breaking changes for older versions? --> **No**
* Does this needs documentation? --> **Yes**

[ZEPPELIN-635]: https://issues.apache.org/jira/browse/ZEPPELIN-635
[ZEPPELIN-689]: https://issues.apache.org/jira/browse/ZEPPELIN-689
